### ISSUE TITLE
Fix SettingsFinder problems

### ISSUE THREAD

**Issue by edreamleo** [leo-editor/leo-editor#419]
Recent revs removed the way-too-wonky SettingsFinder.find_setting, using a default of 2 always for the `which` argument. **Bug**: the default of 2 could create duplicate nodes in myLeoSettings.leo. That would be confusing.

Instead, the code (not the user) should work this way:

1. If the to-be-copied node is found in myLeoSettings.leo, the *only* correct thing to do is to select that node, which includes expanding all ancestors.

2. If the to-be-copied node is *not* found in myLeoSettings.leo, the code should copy *as much as possible without causing duplicates*. For example, if the to-be-copied node's grandfather appears in myLeoSettings.leo, but not the father, then the code should copy only two nodes.

**Comment by edreamleo** [comment]
I am going to put this on hold indefinitely.  This is a minor enhancement to a minor feature.

**Comment by tbnorth** [comment]
This minor feature was a major project to address a huge demand in the user community for easier to modify settings.

Of course it has not survived Leo's foreign code fighting antibodies.

To be fair, the first injury it suffered came from a039305, which could not have been foreseen. The code opens leoSettings.leo (which it may have to open), copies from it to myLeoSettings.leo (which it may have only just created), and then tries to ensure that the target node is selected for the user at the end of all that.  This was working up until a039305.

But clearly the scare tissue sealing off the invader has thickened since then, I'll try and get it back to a working state.  The complaints at the top of this issue: (1) is probably a reflection of the a039305 issue and (2), the excising of the "way too wonky" code doesn't actually prevent the creation of duplicate settings.  This tool was designed to guide a user through the intricacies of Leo's settings system, taking away the user's chance to pick an option runs counter to that intent.

Tangentially, this function would allow a user to tweak appearances in themes, *if* themes use symbolic rather than literal names for things, i.e. @body-bg rather than @solarized-dark1. In fact maybe this is not tangential, moving away from those names probably neutered this function altogether, that's probably the first thing to assess.

**Comment by edreamleo** [comment]
@tbnorth Hehe.  Well, I guess I'm now an official killer T-cell.  I'll look forward to your improvements.  So long as you *never* put up a dialog which a newbie can not possibly understand.

**Comment by tbnorth** [comment]
Heh, maybe we need to start with a common definition of wonky :-)  To me it's always meant jury-rigged, liable to fall apart.  But maybe it also has "policy wonk" connotations, as in too technical to understand? Because while I agree there's a comprehension challenge there with that dialog, the point of this function was to help newbies understand Leo's settings system, so maybe wording could be improved, but not displaying that dialog seems like an admission that Leo's settings are too complicated to explain to anyone without the background to work them out themselves.

So which wonky did you mean?  :-)

**Comment by edreamleo** [comment]
@tbnorth Maybe I meant nerdy instead ;-)

Imo complex dialogs are to be avoided at all costs.  There's never enough context to make the intent clear.  And the longer the explanation, the more scary the dialog is, and the less likely it will do any good.

Besides, copying settings from outline A to outline B should be simple enough even for newbies. The help with automatically finding a setting, in context, *is* a good idea, but why not leave it at that?

**Comment by tbnorth** [comment]
So once again, this is the issue I was bemoaning the other day.  Like Jake's git version reporting, this functionality arose from a lengthy discussion on the mailing list.

See <https://groups.google.com/forum/#!msg/leo-editor/pM8aVJ1D6ao/t98J3vLBFAAJ> and following for endorsements of the solution from you and others.  But now:

 - you reject it as a solution
 - it's broken at an immediate level because of a039305, which, granted, was unforeseeable
 - it's probably more fundamentally broken because of a move away from symbolic names for style elements [the equivalent of show_msg(txt, level='red') vs. show_msg(txt, level=msg_level.urgent)]

It does test resolve a bit for making contributions.

**Comment by edreamleo** [comment]
@tbnorth  Hmm. I definitely did *not* intend to remove the entry that takes people to relevant settings. It looks like I threw out the baby with the bath water.  No wonder you are upset. Let me look into this.

**Comment by tbnorth** [comment]
I wouldn't say upset, I'd say frustrated, as follows:

As it stands in trunk, it's broken.  It just selects a setting in LeoSettings.leo and leaves the user there.  They can edit it, sure, but that's the wrong place to edit settings, "obviously" ;-).

Basically it does nothing useful and would be better removed.  The whole point was to guide the user through the process of copying settings into myLeoSettings, which it no longer does.

Fixing the selection issue (a039305) aside, I guess the argument could be boiled down to a proposal to allow only option 2 from the dialog, despite previous consensus.  So the frustration arises from (a) the absence of any such proposal, rather the code was just hit with a hammer, broken, and left, and (b) the challenge of trying to make useful contributions when something considered useful is so often later redefined as undesirable.

I could try and fix it, but I didn't think it was broken in the first place, and I now longer know what an acceptable solution would look like. Like the git version display and the theme machinery, same pattern.

Anyway, right now I think it pushes the user in the wrong direction, settings editing wise.

The months long discussion that lead up to this solution clearly indicates that newbies do not find managing Leo's settings simple, so while simple things should be done simply, they need to be simple in the first place.  Including to non-coders who want to use Leo for non-code applications.

**Comment by edreamleo** [comment]
> I could try and fix it, but I didn't think it was broken in the first place, and I now longer know what an acceptable solution would look like. Like the git version display and the theme machinery, same pattern.

@tbnorth Ok. I get you are frustrated.

Let me think about what should be done about helping the user with settings. At one time I had a notion about picking between levels 1-4 automatically, without a dialog. The idea was, roughly, to pick the largest valid/useful value. Imo, if an algorithm can't pick an appropriate level, the newbie will have no chance. You have little chance of convincing me otherwise, but feel free to try :-)

> I see the core style-sheet still uses @body-bg etc., so the menu will still work on that and themes that follow that pattern, and I don't think we can compel theme authors to use those names, so it will work for some themes and not others, that's ok.

This sounds like frustration talking. Are you clear that the theme machinery is *exactly* as it was, except for not including comments? This includes the names of all  `@-constants` in the style sheet, so far as I am able to determine. (It's not easy).

I have spent the last three days working almost exclusively with the new stylesheets. Imo, they are a big improvement because it is now easy to localize all changes in the zz sections.

But stylesheets are still problematic.  Undefined Leo settings cause undefined  `@-constants` in the final stylesheet. If there is an easy principle for naming`@-constants` I haven't found it, even after immersing myself in their details.  Fortunately, I'm not going to take a hammer to stylesheets by changing their `@-constants`.

So as not to offend previous contributors, I think I'll relabel the old themes as "monolithic", rather than deprecated.

I don't remember anything about 'git version display'.  Is it related to the git hooks? Afaik, I reverted to the shell scripts.

It is certainly not my intention to frustrate or discourage contributors. Otoh, I'm not going to refrain from attempting to improve contributions.  It's a delicate matter.

I am always willing to consider drastic changes to Leo as a thought experiment. And I always want to remove features that don't do much, but the opportunity to do so rarely presents itself.  Having *all* generators yield copies of positions was a rare success.

**Comment by tbnorth** [comment]
On Sat, 11 Mar 2017 17:12:15 -0800
"Edward K. Ream" <notifications@github.com> wrote:

> Let me think about what should be done about helping the user with
> settings. At one time I had a notion about picking between levels 1-4
> automatically, without a dialog. The idea was, roughly, to pick the
> largest valid/useful value. Imo, if an algorithm can't pick an
> appropriate level, the newbie will have no chance. You have little
> chance of convincing me otherwise, but feel free to try :-)

Of course :-)  My argument would be that the dialog et al wasn't
really intended to be the way to manage settings, it has very little
coverage currently and could never cover everything.  Rather it was
intended to train new users to be able to manage settings themselves,
while at the same time letting them tweak the things people expect to
be able to tweak with no learning curve, like font size.

> > I see the core style-sheet still uses @body-bg etc., so the menu
> > will still work on that and themes that follow that pattern, and I
> > don't think we can compel theme authors to use those names, so it
> > will work for some themes and not others, that's ok.
> 
> This sounds like frustration talking. Are you clear that the theme
> machinery is *exactly* as it was, except for not including comments?
> This includes the names of all  `@-constants` in the style sheet, so
> far as I am able to determine. (It's not easy).

Ok, I took 

"1. @theme base dark. This will be the basis for all future
dark themes.  It will use the existing solarized background colors,
named explicitly.  Another level of indirection in naming would be more
confusing, not less."

to mean `background: @solarized-base0` rather than 
`background: @body-bg`, but if that's not the case, all the better.

> I don't remember anything about 'git version display'.  Is it related
> to the git hooks? Afaik, I reverted to the shell scripts.

I was referring to the reporting of the git version / Leo build version
in the log, git hooks were involved also.

Cheers -Terry

**Comment by edreamleo** [comment]
> My argument would be that the dialog et al wasn't really intended to be the way to manage settings, it has very little coverage currently and could never cover everything. Rather it was intended to train new users to be able to manage settings themselves, while at the same time letting them tweak the things people expect to be able to tweak with no learning curve, like font size.

I think the present "Edit Settings" menu tree is quite useful as it is, for the purpose you state above.  It lets newbies, or anyone else, to jump directly to the desired setting *in the settings tree*, that is, in the proper context. Imo, that's the hard part--no need to guess or search for the desired setting.  Once the user sees the setting, it's easy to copy as much or as little of the tree as the user sees fit.  This will happen naturally, without further explanation, and without any dialog.

So, ymmv, the present Search: Edit Settings menus are a good addition to Leo, just as they are.

**Comment by tbnorth** [comment]
On Sat, 11 Mar 2017 18:34:49 -0800
"Edward K. Ream" <notifications@github.com> wrote:

> > My argument would be that the dialog et al wasn't really intended
> > to be the way to manage settings, it has very little coverage
> > currently and could never cover everything. Rather it was intended
> > to train new users to be able to manage settings themselves, while
> > at the same time letting them tweak the things people expect to be
> > able to tweak with no learning curve, like font size.
> 
> I think the present "Edit Settings" menu tree is quite useful as it
> is, for the purpose you state above.  It lets newbies, or anyone
> else, to jump directly to the desired setting *in the settings tree*,
> that is, in the proper context. Imo, that's the hard part--no need to
> guess or search for the desired setting.  Once the user sees the
> setting, it's easy to copy as much or as little of the tree as the
> user sees fit.  This will happen naturally, without further
> explanation, and without any dialog.

Why?  I do Settings -> Edit Settings -> Body -> Background -> Color

I'm presented with a setting, so I edit it.  Why would I copy it
anywhere, not even knowing there's somewhere to copy it to?

Now I've edited leoSettings.leo, and it will get overwritten
when I update Leo.

At least that's what I'm seeing, the function just jumps the user to
the the setting in leoSettings.leo.  That was not what it did
originally, and seems to me worse than no guidance at all.  At least
reading the docs. or asking on-line might help the user discover
myLeoSettings.leo.  The current behavior just guides the user to an
apparently simple but incorrect solution.

> So, ymmv, the present Search: Edit Settings menus are a good addition
> to Leo, just as they are.

**Comment by edreamleo** [comment]
On Sat, Mar 11, 2017 at 9:48 PM, Terry Brown <notifications@github.com>
wrote:

> On Sat, 11 Mar 2017 18:34:49 -0800
> "Edward K. Ream" <notifications@github.com> wrote:
>
>
> > I think the present "Edit Settings" menu tree is quite useful as it
> ​ ​
> is


> Why?
> ​...Now I've edited leoSettings.leo, and it will get overwritten
> when I update Leo.
>

​No, the user is *taken* to leoSettings.leo.

But I see your point. Trusting the user to know not to alter
leoSettings.leo seems a bit risky.

What is to be done?

Any kind of *info *dialog (Please don't change leoSettings.leo *or* Please
copy settings to myLeoSettings.leo) will quickly become utterly annoying.

Alas, the old dialog was worse. It asks users to supply info (how much to
copy) that they *can't possibly give* because they have to be *in*
leoSettings.leo to know how much they *want* to copy. This is true for
*everybody*, newbies and experts alike.

How about this? Put up an info dialog at most once per session. It won't be
too annoying, and it will warn newbies especially not to modify
leoSettings.leo.

Edward

**Comment by tbnorth** [comment]
On Sun, 12 Mar 2017 05:32:20 -0700
"Edward K. Ream" <notifications@github.com> wrote:

> On Sat, Mar 11, 2017 at 9:48 PM, Terry Brown
> <notifications@github.com> wrote:
> 
> > Why?
> > ​...Now I've edited leoSettings.leo, and it will get overwritten
> > when I update Leo.
> 
> ​No, the user is *taken* to leoSettings.leo.

Right... but this is part of the pattern, the feature's being evaluated
based on its broken state, not its original behavior.

> But I see your point. Trusting the user to know not to alter
> leoSettings.leo seems a bit risky.
> 
> What is to be done?
> 
> Any kind of *info *dialog (Please don't change leoSettings.leo *or*
> Please copy settings to myLeoSettings.leo) will quickly become
> utterly annoying.

But again, the dialog is meant to train users to be able to edit
settings for themselves, not to be used as a long term method for
editing settings - it covers only a small subset.
 
> Alas, the old dialog was worse. It asks users to supply info (how
> much to copy) that they *can't possibly give* because they have to be
> *in* leoSettings.leo to know how much they *want* to copy. This is
> true for *everybody*, newbies and experts alike.

I don't agree.  The dialog offers three levels of copying, single
setting, immediate group, or parent group, with some descriptive text.

> How about this? Put up an info dialog at most once per session. It
> won't be too annoying, and it will warn newbies especially not to
> modify leoSettings.leo.

What about changing the dialog to look like this (see particularly the
new option 1, replacing the old option 4):

The setting '@{specific}' is in the Leo global configuration
file 'leoSettings.leo' and should p̶r̶o̶b̶a̶b̶l̶y̶  be copied to
'myLeoSettings.leo' before editing.
It may make more sense to copy a group or category of settings.

Please enter 1, 2, 3, or 4:
1. just select the setting in 'leoSettings.leo', I'll decide
   how much to copy to myLeoSettings.leo. (Recommended)
2. copy just one setting, '@{specific}'
3. copy the setting group, '{group}'
4. copy the setting whole category, '{category}'

**Comment by edreamleo** [comment]
On Sun, Mar 12, 2017 at 9:17 AM, Terry Brown <notifications@github.com>
wrote:

>
> I don't agree. The dialog offers three levels of copying, single
> setting, immediate group, or parent group, with some descriptive text.
>
​...

> What about changing the dialog to look like this (see particularly the
> new option 1, replacing the old option 4):
>
> The setting '@{specific}' is in the Leo global configuration
> file 'leoSettings.leo' and should p̶r̶o̶b̶a̶b̶l̶y̶ be copied to
> 'myLeoSettings.leo' before editing.
> It may make more sense to copy a group or category of settings.
>
> Please enter 1, 2, 3, or 4:
> 1. just select the setting in 'leoSettings.leo', I'll decide
> how much to copy to myLeoSettings.leo. (Recommended)
> 2. copy just one setting, '@{specific}'
> 3. copy the setting group, '{group}'
> 4. copy the setting whole category, '{category}'


Here is my understanding of the situation:

You think it is dangerous for newbies to open leoSettings.leo. They might
be tempted to change leoSettings.leo instead of myLeoSettings.leo.

All commands in the Settings: Edit Settings menu copy settings from
leoSettings.leo to myLeoSettings.leo, if the setting does not already exist
in myLeoSettings.leo. The commands then select the (possibly copied) node
in myLeoSettings.leo.

This operation applies to both the original code and the present (interim)
code. The original commands put up a dialog, asking the user how many
ancestor nodes to copy.  The present version copies a settings node and its
parent.

We are presently discussing at least four things:

1. Whether to open leoSettings.leo if the desired setting does not exist in
myLeoSettings.leo.

2. Whether to copy nodes automatically from leoSettings.leo to
myLeoSettings.leo if they do not already exist in myLeoSettings.leo.

3. If copying is to happen, whether to use a dialog.

4. If a dialog is to be used, what its exact form should be.

Please correct me if what I have just said is inaccurate in any way.

Edward

**Comment by tbnorth** [comment]
On Sun, 12 Mar 2017 11:56:12 -0700
"Edward K. Ream" <notifications@github.com> wrote:

> Here is my understanding of the situation:
> 
> You think it is dangerous for newbies to open leoSettings.leo. They
> might be tempted to change leoSettings.leo instead of
> myLeoSettings.leo.
> 
> All commands in the Settings: Edit Settings menu copy settings from
> leoSettings.leo to myLeoSettings.leo, if the setting does not already
> exist in myLeoSettings.leo. The commands then select the (possibly
> copied) node in myLeoSettings.leo.

That's not what I see with

Leo 5.4, build 20170312083401, Sun Mar 12 08:34:01 CDT 2017
Git repo info: branch = master, commit = 92b86364d6db
Python 2.7.12, PyQt version 5.5.1
linux2

running Leo in vanilla mode with the
HOME=/tmp python launchLeo.py trick.

myLeoSettings.leo is opened, but nothing is copied to it.  The user is
left looking at the node selected in leoSettings.leo.

What your describing sounds more like the way the feature worked
originally.  I'm not sure if the original code closed leoSettings.leo
after it had copied the nodes, it certainly left the user looking at
myLeoSettings.leo.  And again, no one can be blamed for the unforeseen
side effects of a039305.

> This operation applies to both the original code and the present
> (interim) code. The original commands put up a dialog, asking the
> user how many ancestor nodes to copy.  The present version copies a
> settings node and its parent.
>
> We are presently discussing at least four things:
> 
> 1. Whether to open leoSettings.leo if the desired setting does not
> exist in myLeoSettings.leo.

Which would be option (1) (Recommended) in my proposed revision of the
dialog.

> 2. Whether to copy nodes automatically from leoSettings.leo to
> myLeoSettings.leo if they do not already exist in myLeoSettings.leo.

I'd argue automatic copying with no dialog loses the opportunity to
teach the user how settings work.
 
> 3. If copying is to happen, whether to use a dialog.
> 
> 4. If a dialog is to be used, what its exact form should be.
> 
> Please correct me if what I have just said is inaccurate in any way.

Apart from the difference is observed current behavior of the feature,
I think that covers it.

Cheers -Terry

> Edward
> 
>

**Comment by edreamleo** [comment]
On Sun, Mar 12, 2017 at 2:31 PM, Terry Brown <notifications@github.com>
wrote:

​I have two choices, neither of which please me completely:

1. Make a pronouncement as a supposedly benevolent dictator.

In that case, the commands would simply find a setting in leoSettings.leo.

2. Leave the commands​ to Terry.

After some thought I have decided to follow option 2.

We have enough dictatorships in this world. None of them are benevolent.
More positively, you, Terry, have argued at length for your point of view.
I am not persuaded, but I want to reward your diligence.

I will not write code to determine context automatically.  Terry, you can
do that, or simply offer 3 or 4 static options in the dialog.

This discussion has been useful, but it is time to move on.

Edward

P.S. As a dictator, I would issue the following rulings.  To repeat, I am
not going to do that, but I want to emphasize, one last time, the reasons
for my displeasure. Furthermore, if problems do arise later, I may choose
to revise my present decision.

Ruling 1.  There is indeed a danger that newbies may be tempted to change
leoSettings.leo instead of myLeoSettings.leo. However, this danger is not
serious.  Newbies can open leoSettings.leo in many ways.  There is ample
warnings in various places, including leoSettings.leo, that users should
alter myLeoSettings.leo instead.

Ruling 2: The commands in the Settings: Edit Settings menu are useful, both
for newbies and experts.  These commands are incomplete, but that does not
decrease their utility.  They could be expanded if desired. They are not
likely to become obsolete because the settings in leoSettings.leo change
only if all references to them in Leo's core disappear.

Ruling 3: The commands under discussion should open myLeoSettings.leo if
the setting exists there.  Otherwise, the command should open
leoSettings.leo. In either case, the desired node should be selected. This
operation is useful for newbies and experts alike.  Users can then easily
select whatever part of leoSettings they like, and copy that tree to
myLeoSettings.leo.

EKR

**Comment by edreamleo** [comment]
I am going to leave this issue open.  Terry, please make your desired changes quickly.  I would like to release 5.5b1 in a few days, with this issue firmly closed.

**Comment by tbnorth** [comment]
On Sun, 12 Mar 2017 14:17:54 -0700
"Edward K. Ream" <notifications@github.com> wrote:

> I am going to leave this issue open.  Terry, please make your desired
> changes quickly.  I would like to release 5.5b1 in a few days, with
> this issue firmly closed.

I will do what I can but I'm not sure how easy this will be to
resolve.  For me, time to work on Leo is a very scarce resource.  The
side effects of a039305 may or may not be the only challenge to
restoring the function's original behavior.

I will try and implement the dialog described at the end of
https://github.com/leo-editor/leo-editor/issues/419#issuecomment-285947526
with perhaps slightly more didactic wording.

**Comment by edreamleo** [comment]
On Sun, Mar 12, 2017 at 8:43 PM, Terry Brown <notifications@github.com>
wrote:

> On Sun, 12 Mar 2017 14:17:54 -0700
> "Edward K. Ream" <notifications@github.com> wrote:
>
> > I am going to leave this issue open. Terry, please make your desired
> > changes quickly. I would like to release 5.5b1 in a few days, with
> > this issue firmly closed.
>
> I will do what I can but I'm not sure how easy this will be to
> resolve. For me, time to work on Leo is a very scarce resource.


​But not for me :-) When I awoke I realized the way forward is to move the
SettingsFinder class into a settings_finder plugin.  I'll do that and
restore your original code.

Edward

**Comment by edreamleo** [comment]
Rev eb52ec4 completes work on the settings_finder plugin.  There were conflicting commits concerning the dialog message. The latest rev restores Terry's recent changes, but Terry, please verify that everything as you like.

Afaik, the plugin works exactly as before, except that the plugin creates the Edit Settings menu at the end of the Settings menu.

Terry, feel free to make any changes you like to the plugin.  It's yours.  I even preserved gnx's.

This will conclude work on this issue, from my point of view.

**Comment by tbnorth** [comment]
@edreamleo The function was broken again, I fixed it again (1cd3ba4).  I assume the 'g' in [this line](https://github.com/leo-editor/leo-editor/commit/eb52ec412d4d7032c30299c576fceafeef2ac162#diff-953691c2ec1750dee988e32e2befbafeL148) was a typo., I removed it.

The plug in is not mentioned in leoSettings.leo.  If the intent is that it not be enabled by default, then I consider it effectively removed from the code base. A function intended to help new users manage settings is pointless as a disabled by default plug in with no path to discovery, not to mention the catch-22 of needing to edit settings to enable it.

I have a fix for the bug (duplicate settings) at the top of this issue, really don't think it's worth applying the the plug in is not to be enabled by default though.  Let me know.
